### PR TITLE
Initial GET and DELETE Operations

### PR DESCRIPTION
**Pull request by mprorock**
The bare minimum get and delete operations for credentials and presentations - putting this in for feedback and once merged we can extend the pattern as required.  in particular - i think we will want allow a PUT to update and / or replace items by ID

NOTE: Depends on pr #270 and should not be merged until that is resolved

### CODE REVIEW

**Comment by msporny** [comment]
> NOTE: Depends on pr #270 and should not be merged until that is resolved

Marking 'do not merge' temporarily so no one accidentally merges this before #270.

**Comment by mprorock** [comment]
@msporny i am inclined to set "id" for getters to be UUID or a URI - does this make sense to you?

**Comment by msporny** [comment]
> @msporny i am inclined to set "id" for getters to be UUID or a URI - does this make sense to you?

I'm fine to start there. I expect that we'll quickly move towards something that looks more like 'transaction-id' in time. Every time we try to pick UUID, it turns into "UUID and multi-base 128 bit value", which then turns into "but I want to encode JWT-like objects in the ID, so give me 1024 URL-safe characters to play around with".

I get why we'd want URIs for GET... shudder a bit at having to URIEncode those. No strong feelings, but again, feels like "id"s are just going to turn into "1024 URL-safe characters" over time. Just my $0.02 for now.

**Comment by mprorock** [comment]
> just going to turn into "1024 URL-safe characters" over time. Just my $0.02 for now.

yeah - i suspect you are right for top level API and then i also expect profiles to tighten that down to their specific requirements - e.g. i would bet the traceability goes with just UUID for a variety of reasons

**Review by OR13 — APPROVED** [review]

**Comment by peacekeeper** [comment]
I think this is useful, but GET and DELETE operations should probably be optional. There may be implementations that can issue VCs but then don't remember them.

**Comment by msporny** [comment]
This needs to be refactored on top of main. I suggest cherry picking the four commits instead of rebasing -- I cleaned up the history to remove the add/removal of package-lock.json as well as the auto-generated files in the commit history.

**Comment by OR13** [comment]
There are conflicts

**Comment by msporny** [comment]
> Conflicts should now be resolved

Still showing conflits with rebase and merge. I can squash-merge if you'd like, but would prefer to keep your history.

**Comment by OR13** [comment]
Screen Shot 2022-03-29 at 8 58 08 AM

The other code owners have not reviewed.

File: holder.yml
Comment (msporny) [inline holder.yml:24]: Hrm, disagree w/ this only being able to be a UUID -- but we can deal w/ that in another PR. Just noting my objection to this one design choice.

File: holder.yml
Comment (msporny) [inline holder.yml:41]: Har har :) -- presuming we're going to eventually remove this... because if we don't, we'll have to test for it.

File: holder.yml
Comment (msporny) [inline holder.yml:59]: This should probably say "Credential marked for deletion". Looks like the proposal here is that this call is sometimes implemented as eventually consistent.

**Review by msporny — APPROVED** [review]
Ok to merge as a work in progress. I've noted a few things that are concerning with the PR, but good enough for now.

**Comment by msporny** [comment]
This is what I see: 

image

File: holder.yml
Comment (OR13) [reply holder.yml:24]: @mprorock we can take this change in our profile and leave this repo with `type` string where `<script> alert(1);</script>` is a spec legal identifier.

Comment (OR13) [inline holder.yml:24]:
```suggestion
```

Comment (OR13) [reply holder.yml:24]: Lets make sure to allow xss / command injection in our identifiers.

File: holder.yml
Comment (OR13) [inline holder.yml:37]: are we going to define authorization? and test it?

File: holder.yml
Comment (OR13) [inline holder.yml:85]:
```suggestion
                - "verifiable-credentials"
```

File: holder.yml
Comment (msporny) [reply holder.yml:24]: > leave this repo with `type` string where `<script> alert(1);</script>` is a spec legal identifier.

Or, you know, do something more sane and re-use our existing transactionId/exchangeId regexes, which don't allow the attack you mention above:

https://github.com/w3c-ccg/vc-api/blob/main/components/parameters/path/TransactionId.yml

File: holder.yml
Comment (msporny) [reply holder.yml:37]: > are we going to define authorization? and test it?

The test suites will need to have some authz mechanism in them because the production software that all implementers are implementing are all (AFAIK) requiring authz. If we have interop on the test suites, the spec will have to reflect reality and document how authz is being performed in an interoperable fashion.

File: holder.yml
Comment (msporny) [reply holder.yml:85]: Why kebab-case instead of camelCase? We're using camelCase everywhere else, no?

**Review by msporny — CHANGES_REQUESTED** [review]
Hrm, just spotted the two massive auto-generated files. We have Github Actions to auto-generate files, we should not commit autogenerated files.

File: holder.yml
Comment (OR13) [reply holder.yml:85]: its a query string param,  and its pluralized... so i figured it would be better to not have it look like `verifiableCredential` which shows up in VPs, but I am not attached to anything other than fixing the lowercase concatenation issue.

**Review by OR13 — CHANGES_REQUESTED** [review]
remove auto generated files

File: holder.yml
Comment (mprorock) [reply holder.yml]: +1 manu - added an `ObjectId` that matches that transaction id

**Comment by mprorock** [comment]
> Hrm, just spotted the two massive auto-generated files. We have Github Actions to auto-generate files, we should not commit autogenerated files.

that should be fixed now - thanks for catching that

**Review by OR13 — APPROVED** [review]

**Comment by msporny** [comment]
> ahh thats because you are using rebase... you should stop using rebase since it makes it harder for non technical git users to contribute ; )

When there are non-technical Github users involved, I use squash. 

This is not one of those cases :), @mprorock knows what he's doing. I consider it bad form to squash merge contributor's commits if it looks like they've gone to the trouble of creating a clean history. In those cases, I ask them if they're ok with it before doing it.

File: holder.yml
Comment (msporny) [inline holder.yml:75]: These options seem a bit arbitrary... fine for now, but we need to define exactly what should be happening when each option is provided... for example, is this really checking the `type` field, or is it doing something else -- my guess is the latter... while the former might be more useful?

**Review by msporny — APPROVED** [review]
LGTM. @mprorock, do I have your permission to squash-merge this PR?

**Comment by mprorock** [comment]
> LGTM. @mprorock, do I have your permission to squash-merge this PR? Or, would you like to rebase so I can rebase-merge? Note: If you rebase, please just make it so that the auto-generated files don't show up in history at all... I had to go in and surgically remove auto-gen'd files from history the last time.

personally I am a squash and merge fan

**Pull request merged**
